src/ledger/LedgerState.cpp

```diff
@@ -0,0 +1,1311 @@
+#include "xdrpp/marshal.h"
+#include <soci.h>
+
+namespace stellar
+{
+
+bool
+isBetterOffer(LedgerEntry const& lhsEntry, LedgerEntry const& rhsEntry)
+{
+    auto const& lhs = lhsEntry.data.offer();
+    auto const& rhs = rhsEntry.data.offer();
+
+    assert(lhs.buying == rhs.buying);
+    assert(lhs.selling == rhs.selling);
+
+    double lhsPrice = double(lhs.price.n) / double(lhs.price.d);
+    double rhsPrice = double(rhs.price.n) / double(rhs.price.d);
+    if (lhsPrice < rhsPrice)
+    {
+        return true;
+    }
+    else if (lhsPrice == rhsPrice)
+    {
+        return lhs.offerID < rhs.offerID;
+    }
+    else
+    {
+        return false;
+    }
+}
+
+// Implementation of AbstractLedgerState --------------------------------------
+AbstractLedgerStateParent::Identifier
+AbstractLedgerStateParent::getIdentifier() const
+{
+    return Identifier();
+}
+
+// Implementation of LedgerState ----------------------------------------------
+LedgerState::LedgerState(AbstractLedgerStateParent& parent,
+                         bool shouldUpdateLastModified)
+    : mImpl(std::make_unique<Impl>(*this, parent, shouldUpdateLastModified))
+{
+}
+
+LedgerState::LedgerState(LedgerState& parent, bool shouldUpdateLastModified)
+    : LedgerState((AbstractLedgerStateParent&)parent, shouldUpdateLastModified)
+{
+}
+
+LedgerState::Impl::Impl(LedgerState& self, AbstractLedgerStateParent& parent,
+                        bool shouldUpdateLastModified)
+    : mParent(parent), mChild(nullptr), mHeader(mParent.getHeader())
+    , mShouldUpdateLastModified(shouldUpdateLastModified)
+    , mIsSealed(false)
+{
+    mParent.addChild(self);
+}
+
+LedgerState::~LedgerState()
+{
+    if (mImpl)
+    {
+        rollback();
+    }
+}
+
+void
+LedgerState::addChild(AbstractLedgerState& child)
+{
+    mImpl->addChild(child);
+}
+
+void
+LedgerState::Impl::addChild(AbstractLedgerState& child)
+{
+    checkNotSealed();
+    checkNoChild();
+
+    mChild = &child;
+    mActive.clear();
+    mActiveHeader.reset();
+}
+
+void
+LedgerState::Impl::checkNoChild() const
+{
+    if (mChild)
+    {
+        throw std::runtime_error("LedgerState has child");
+    }
+}
+
+void
+LedgerState::Impl::checkNotSealed() const
+{
+    if (mIsSealed)
+    {
+        throw std::runtime_error("LedgerState is sealed");
+    }
+}
+
+void
+LedgerState::commit()
+{
+    mImpl->commit(getIdentifier());
+    mImpl.reset();
+}
+
+void
+LedgerState::Impl::commit(Identifier id)
+{
+    updateLastModified(); // Invokes checkNoChild
+
+    mActive.clear();
+    mActiveHeader.reset();
+    mParent.commitChild(id);
+}
+
+void
+LedgerState::commitChild(Identifier id)
+{
+    mImpl->commitChild(*this);
+}
+
+void
+LedgerState::Impl::commitChild(LedgerState& self)
+{
+    auto entries = mChild->getEntries();
+    for (auto const& kv : entries)
+    {
+        auto const& key = kv.first;
+        if (kv.second)
+        {
+            mEntry[key] = std::make_shared<LedgerEntry>(*kv.second);
```

Comment:
maybe not exactly like this (it would require special casing exception handling), my point is more that it's OK to have destructive semantics with the child, so if you were "moving" entries one by one into the parent, we can avoid the copying